src/common/transformations/src/transformations/transpose_sinking/ts_unsqueeze.cpp

```diff
@@ -0,0 +1,243 @@
+// Copyright (C) 2018-2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "transformations/transpose_sinking/ts_unsqueeze.hpp"
+
+#include <memory>
+#include <vector>
+
+#include "itt.hpp"
+#include "openvino/core/validation_util.hpp"
+#include "openvino/opsets/opset10.hpp"
+#include "openvino/pass/pattern/op/wrap_type.hpp"
+#include "transformations/rt_info/transpose_sinking_attr.hpp"
+#include "transformations/transpose_sinking/ts_utils.hpp"
+#include "transformations/utils/utils.hpp"
+
+using namespace ov;
+using namespace opset10;
+using namespace ov::pass::pattern;
+using namespace ov::pass::transpose_sinking;
+using namespace ov::pass::transpose_sinking::utils;
+
+namespace {
+
+/**
+ * @brief Checks that Reshape operation is equal to Unsqueeze:
+ * Only 1 dims are inserted, all other dims must be the same.
+ * Converts these 1 dims to axes format.
+ * @arg reshape Reshape operation.
+ * @arg reshape_to_shape 2nd input to Reshape op as a constant.
+ * @arg result_axes contains axes which will be unsqueezed.
+ */
+bool shape_to_unsqueeze_axes(const std::shared_ptr<Node>& reshape,
+                             const std::shared_ptr<Constant>& reshape_to_shape,
+                             std::vector<size_t>& result_axes) {
+    result_axes.clear();
+    auto reduction_axes_values = reshape_to_shape->cast_vector<int64_t>();
+    // supported the case if Reshape is equal to Unsqueeze
+    const auto& new_shape = reduction_axes_values;
+    const auto& input_pshape = reshape->get_input_partial_shape(0);
+    // todo: support dynamic case
+    if (input_pshape.is_dynamic()) {
+        return false;
+    }
+
+    const auto input_shape = input_pshape.to_shape();
+    if (new_shape.size() > input_shape.size()) {
+        size_t j = 0;
+        for (size_t i = 0; i < new_shape.size(); ++i) {
+            if (j < input_shape.size() && static_cast<int64_t>(input_shape[j]) == new_shape[i]) {
+                j++;
+            } else if (new_shape[i] != 1) {
+                return false;
+            } else {
+                result_axes.push_back(i);
+            }
+        }
+        if (j != input_shape.size()) {
+            // not all input_shape values are in new_shape
+            return false;
+        }
+    } else {
+        // another reshape type, not Unsqueeze
+        // todo: move this checks in the pattern
+        return false;
+    }
+    return true;
+}
+
+/**
+ * @brief Converts unsqueeze_axes to actual shape (2nd input) for Reshape operation
+ * using the shape of the 1st input to Reshape.
+ * @arg input_node 1st input to Reshape op.
+ * @arg unsqueeze_axes In case of Reshape op is equal to Unsqueeze, these axes indicate the places where 1 dims have
+ * to be inserted.
```

Comment:
```suggestion
 * @arg to_shape shape resulted with Unsqueeze operation.
```

Reply:
do we have such functions used for shape inference for these operations? May be we can re-use them and they are tested well?